<div class="alert alert-block alert-info">
Singapore Management University<br>
CS105 Statistical Thinking for Data Science, 2023/24 Term 2
</div>

# CS105 Group Project Submission (Part I)

-----
Provide your team details, including section, team number, team members, and the name of the dataset. 
Complete all of the following sections. For any part requiring code to derive your answers, please create a code cell immediately below your response and run the code.
To edit any markdown cell, double click the cell; after editing, execute the markdown cell to collapse it.
<br>
-----

## Declaration

<span style="color:red">By submitting this notebook, we declare that **no part of this submission is generated by any AI tool**. We understand that AI-generated submissions will be considered as plagiarism, and just like other plagirisum cases, disciplinary actions will be imposed.</span>

#### Section: G1
#### Team: Group 1
#### Members:
1. CHONG SHIH RUI ZACHARY
2. YANG ZHONG HNG REUBEN
3. KEAGAN PAUL POULOSE
4. LUCAS SONG KANGZHENG
5. 

#### Dataset: Wine dataset

## Part I: Exploratory Data Analysis (EDA) [8% of final grade]

In [79]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

In [80]:
df = pd.read_csv('wine.csv')

In [81]:
df.head()

ID  Fixed_Acidity  Volatile_Acidity  Citric_Acid Residual_Sugar  Chlorides  \
0   1            7.0              0.27         0.36           High      0.045   
1   2            6.3              0.30         0.34            Low      0.049   
2   3            8.1              0.28         0.40            Low      0.050   
3   4            7.2              0.23         0.32       Moderate      0.058   
4   5            7.2              0.23         0.32       Moderate      0.058   

   Free_SO2  Total_SO2    Density      pH  Sulphates Alcohol  Quality  
0      45.0      170.0      >1.00  Acidic       0.45      8%        6  
1      14.0      132.0  0.99-1.00  Acidic       0.49      9%        6  
2      30.0       97.0  0.99-1.00  Acidic       0.44     10%        6  
3      47.0      186.0  0.99-1.00  Acidic       0.40      9%        6  
4      47.0      186.0  0.99-1.00  Acidic       0.40      9%        6

In [82]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4898 entries, 0 to 4897
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   ID                4898 non-null   int64  
 1   Fixed_Acidity     4680 non-null   float64
 2   Volatile_Acidity  4848 non-null   float64
 3   Citric_Acid       4898 non-null   float64
 4   Residual_Sugar    4898 non-null   object 
 5   Chlorides         4898 non-null   float64
 6   Free_SO2          4898 non-null   float64
 7   Total_SO2         4898 non-null   float64
 8   Density           4674 non-null   object 
 9   pH                4898 non-null   object 
 10  Sulphates         4898 non-null   float64
 11  Alcohol           4898 non-null   object 
 12  Quality           4898 non-null   int64  
dtypes: float64(7), int64(2), object(4)
memory usage: 497.6+ KB


In [83]:
df["Residual_Sugar"].unique()

array(['High', 'Low', 'Moderate'], dtype=object)

### 1. Overview of dataset [15% of Part I]

**a.** Summarise the background of the dataset [limited to 50 words]

**This dataset gives the quality of wine.
Each row is an observation and represents a type of wine, and along each row, you find values that describe certain attributes about the wine. In particular, the variable Quality indicates whether the quality of the wine leans more towards good or bad.**

**b.** State the size of the dataset

In [84]:
df.size #Size of dataset

63674

**As shown above, the size of the Wine dataset is 63674** 

**c.** For each variable, describe what it represents and its data type (numerical or categorical)

|Column name|Description|Data type|
|---|---|---|
|ID| Unique Id of a wine | Categorical |
|Fixed_Acidity| Amount of Tartaric acid, measured in g/dm3 | Numerical |
|Volatile_Acidity| Amount of Acetic acid, measured in g/dm3  | Numerical |
|Citric_Acid| Amount measured in g/dm3 | Numerical |
|Residual_Sugar| Level of residual sugar content | Categorical (“High”, “Moderate”, “Low”) |
|Chlorides| Amount of Sodium chloride, measured in g/dm3 | Numerical |
|Free_SO2| Amount measured in mg/dm3 | Numerical |
|Total_SO2| Amount measured in mg/dm3  | Numerical |
|Density| Amount measured in g/dm3 | Categorical (“<0.99”, “0.99-1.00”, “>1.00”) |
|pH| The measure of acidity of the wine | Categorical (“Neutral”, “Acidic”, “Highly Acidic”)| 
|Sulphates|Potassium sulphate, measured in g/dm3 | Numerical |
|Alcohol| Level of alcohol content | Categorical (8% through 14% in discrete levels)|
|Quality| Score based on sensory data | Categorical (0 through 10 (0 is bad, 10 is good))|

### 2. Data pre-processing [35% of Part I]

**a.** For each variable, determine the percentage of missing data. For any column with missing data, describe how you resolve the issue. Clearly state any assumption you made.

In [85]:
n_rows, n_cols = df.shape #Number of rows and columns in wine dataset
(1 - df.count() / n_rows)*100 #Percentage of missing data

ID                  0.000000
Fixed_Acidity       4.450796
Volatile_Acidity    1.020825
Citric_Acid         0.000000
Residual_Sugar      0.000000
Chlorides           0.000000
Free_SO2            0.000000
Total_SO2           0.000000
Density             4.573295
pH                  0.000000
Sulphates           0.000000
Alcohol             0.000000
Quality             0.000000
dtype: float64

|Column name|Percentage of missing data|How to resolve|
|---|---|---|
|ID| 0% | NA |
|Fixed_Acidity| 4.450796% | Less than 5%, so we should impute |
|Volatile_Acidity| 1.020825%  | Less than 5%, so we should impute |
|Citric_Acid| 0% | NA |
|Residual_Sugar| 0% | NA |
|Chlorides| 0% | NA |
|Free_SO2| 0% | NA |
|Total_SO2| 0%  | NA |
|Density| 4.573295% | Less than 5%, so we should impute |
|pH| 0% | NA | 
|Sulphates|0%| NA |
|Alcohol| 0% | NA |
|Quality| 0% | NA |

**Whether we use mean/mode for imputation, probably should explain why**

**b.** For each variable, identify outliers (if any) and describe how you resolve the issue. Clearly state any assumption you made.

**Response.** 

**c.** For categorical variables, perform the necessary encoding.

In [86]:
#Unique values of Residual_Sugar
df["Residual_Sugar"].unique() 

array(['High', 'Low', 'Moderate'], dtype=object)

In [87]:
#Unique values of Density
df["Density"].unique()

array(['>1.00', '0.99-1.00', nan, '<0.99'], dtype=object)

In [88]:
#Unique values of pH
df["pH"].unique()

array(['Acidic', 'Highly Acidic', 'Neutral'], dtype=object)

In [89]:
df = pd.read_csv('wine.csv')

In [90]:
from sklearn.preprocessing import OrdinalEncoder

enc = OrdinalEncoder()

#Encoding Residual_Sugar
X = [['Low'],['Moderate'],['High']]
enc.fit(X)
df[["Residual_Sugar_enc"]] = enc.transform(df[["Residual_Sugar"]])

#Encoding Density
#Y = [['>1.00'],['0.99-1.00'],['<0.99']]
#enc.fit(Y)
#df[["Density_enc"]] = enc.transform(df[["Density"]])

#Encoding pH
Z = [['Neutral'],['Acidic'],['Highly Acidic']]
enc.fit(Z)
df[["pH_enc"]] = enc.transform(df[["pH"]])

C:\Users\User\anaconda3\Lib\site-packages\sklearn\base.py:457: UserWarning: X has feature names, but OrdinalEncoder was fitted without feature names
  warnings.warn(
C:\Users\User\anaconda3\Lib\site-packages\sklearn\base.py:457: UserWarning: X has feature names, but OrdinalEncoder was fitted without feature names
  warnings.warn(


In [91]:
df.head()

ID  Fixed_Acidity  Volatile_Acidity  Citric_Acid Residual_Sugar  Chlorides  \
0   1            7.0              0.27         0.36           High      0.045   
1   2            6.3              0.30         0.34            Low      0.049   
2   3            8.1              0.28         0.40            Low      0.050   
3   4            7.2              0.23         0.32       Moderate      0.058   
4   5            7.2              0.23         0.32       Moderate      0.058   

   Free_SO2  Total_SO2    Density      pH  Sulphates Alcohol  Quality  \
0      45.0      170.0      >1.00  Acidic       0.45      8%        6   
1      14.0      132.0  0.99-1.00  Acidic       0.49      9%        6   
2      30.0       97.0  0.99-1.00  Acidic       0.44     10%        6   
3      47.0      186.0  0.99-1.00  Acidic       0.40      9%        6   
4      47.0      186.0  0.99-1.00  Acidic       0.40      9%        6   

   Residual_Sugar_enc  pH_enc  
0                 0.0     0.0  
1                 1.0     0.0  
2                 1.0     0.0  
3                 2.0     0.0  
4                 2.0     0.0

### 3.	Exploratory analysis and visualization [50% of Part I]

**a.** For each variable, provide relevant summary statistics

**Response.** 

**b.** For each variable, provide an appropriate visualisation depicting the distribution of its values, and summarize any key observation(s) you made.

**Response.** 

**c.** Perform bi-variate analysis on the variables. You do not need to present the analysis of every pair of variables; only focus on the pairs you believe are worth investigating and explain. For each pair, describe the relationship between the two variables. Use appropriate statistical methods and/or visualizations.

**Response.** 